### PR TITLE
feat(payments-paypal): create PaypalService createBillingAgreement method

### PR DESCRIPTION
**Pull request by sardesam**
## Because

- Part of M3a.
- We want `createBillingAgreement` added to `PaypalService`.

## This pull request

- Adds `createBillingAgreement` to `PaypalService`.
- Adds `createBillingAgreement` to `PaypalManager`.
- Adds `NVPCreateBillingAgreementResponseFactory` to `factories` in `payments-paypal`.
- Adds `CurrencyManager` library.
- Adds/updates all applicable factories and tests.

## Issue that this pull request solves

Closes FXA-8935

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

### CODE REVIEW

File: libs/payments/paypal/src/lib/paypal.service.ts [hunk 1]
```diff
@@ -3,9 +3,57 @@
...
+
+    const agreementDetails = await this.paypalManager.getBillingAgreement(
+      billingAgreementId
+    );
+
+    const country = agreementDetails.countryCode;
+
+    if (
+      this.currencyManager.isCurrencyCompatibleWithCountry(currency, country)
+    ) {
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.ts:46]: I think we'd probably short circuit here and throw an error, since we wouldn't want to proceed if currency is incompatible.

File: libs/payments/currency/src/lib/currency.manager.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    // Must be 10 digits or less
+    if (amountInCents.toString().length > 10) {
+      throw new AmountExceedsDigitLimitError(amountInCents);
+    }
+    // Left pad with zeros if necessary, so we always get a minimum of 0.01.
+    const amountAsString = String(amountInCents).padStart(3, '0');
+    const dollars = amountAsString.slice(0, -2);
+    const cents = amountAsString.slice(-2);
+    return `${dollars}.${cents}`;
+  }
```
Comment (julianpoy) [inline libs/payments/currency/src/lib/currency.manager.ts:75]: This does seem like something we may want to move to the paypal library rather than having this in the currency manager. Perhaps into the paypal manager?

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 1 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts:46]: What's the best way to do that? I put the throw in `this.currencyManager.isCurrencyCompatibleWithCountry(currency, country) for the mismatch`.

File: libs/payments/paypal/src/lib/paypal.service.ts [hunk 3]
```diff
@@ -3,9 +3,57 @@
...
+  async createBillingAgreement(options: {
+    uid: string;
+    token: string;
+    currency: string;
+  }) {
+    const { uid, currency } = options;
+
+    const billingAgreement = await this.paypalClient.createBillingAgreement(
+      options
+    );
```
[the remark this answers is not in the document]
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts:34]: Ah, so I'm never actually hitting the client directly? Does that mean I'll have to add this `createBillingAgreement` to `PaypalManager`? Double checking as the description in the ticket said:

> Call paypal client createBillingAgreement

File: libs/payments/currency/src/lib/currency.manager.ts (hunk 2 above)
Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.ts:75]: Agreed. It was part of `CurrencyHelper` in payments and isn't actually used at all for this ticket. Since I moved the whole helper over, I thought I might as well update it and add tests. It is pretty PayPal-specific. And It still seems relevant (https://developer.paypal.com/docs/api/payments/v1/#definition-amount). Will update!

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (julianpoy) [reply libs/payments/paypal/src/lib/paypal.service.ts:34]: Oops...

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 1 above)
Comment (julianpoy) [reply libs/payments/paypal/src/lib/paypal.service.ts:46]: Ah, a couple of things here then:
1. A method that's named starting with `is` implies that the method will return `true` or `false` - I wouldn't expect a method with `is` to throw on false.
2. If you want to have isCurrencyCompatibleWithCountry throw, I'd rename it to assertCurrencyCompatibleWithCountry - that's the usual naming to imply that it'll throw an error if false.
3. As it stands, since the method currently functions like an `assert`, you don't need it in the `if` block at all, since the `if` block doesn't really do anything - just have the call to isCurrencyCompatibleWithCountry before the next line with no if block and it'll exit early due to the thrown error.

Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts:46]: Ah! Got it! Thank you!

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts:34]: Updated to add method to PaypalManager (figured I'd go ahead and do it while I'm at it). Still have to add tests but just an FYI! 🙂

File: libs/payments/currency/src/lib/currency.manager.ts (hunk 2 above)
Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.ts]: Moved over to `PayPalManager` along with error to `paypal.error.ts`. Same as above, still have to do some tweaks and finish moving tests over.

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 1 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts]: Done!

File: libs/payments/currency/src/lib/currency.manager.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+   * valid currencies and countries as listed in constants
+   *
+   * @param currency Currency of customer
+   * @param country Country of customer
+   * @returns True if currency is compatible with country, else throws error
+   */
+  assertCurrencyCompatibleWithCountry(
+    currency: string | null | undefined,
+    country: string | null | undefined
+  ): boolean {
```
Comment (julianpoy) [inline libs/payments/currency/src/lib/currency.manager.ts:33]: The return type for an `assert`-style function would be void.

Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.ts]: Updated along with test!

File: libs/payments/paypal/src/lib/paypal.manager.ts [hunk 5]
```diff
@@ -35,6 +37,19 @@ export class PayPalManager {
...
   }
 
+  /**
+   * Create billing agreement using the ExpressCheckout token.
+   *
+   * If the call to PayPal fails, a PayPalClientError will be thrown.
+   *
+   */
+  async createBillingAgreement(
+    options: CreateBillingAgreementOptions
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.manager.ts:47]: This should take a string with token as an argument so that it matches your jsdoc

File: libs/payments/paypal/src/lib/paypal.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      paypalCustomerManager
+    );
+
+    currencyManager = new CurrencyManager();
+
+    paypalService = new PayPalService(
+      paypalManager,
+      paypalCustomerManager,
+      currencyManager
+    );
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.spec.ts:57]: This should all be constructed with NestJS's test dependency injection tooling.

https://docs.nestjs.com/fundamentals/testing#testing-utilities

See an example of it's usage here:

https://github.com/mozilla/fxa/blob/3354f25b44b06326bfa368950d8909fdfd4262d4/libs/shared/contentful/src/lib/contentful.manager.spec.ts#L46

and here:

https://github.com/mozilla/fxa/blob/3354f25b44b06326bfa368950d8909fdfd4262d4/libs/payments/stripe/src/lib/stripe.service.spec.ts#L12

**Comment by julianpoy** [comment]
There are a lot of files changed in this PR that seem unrelated to the PR title

File: packages/fxa-payments-server/server/lib/server.js [hunk 7]
```diff
@@ -30,12 +30,6 @@ module.exports = () => {
   const { cors, routing } = require('fxa-shared/express').express();
   const { v4: uuid } = require('uuid');
 
-  const {
-    tagCriticalEvent,
-    buildSentryConfig,
-    tagFxaName,
-  } = require('fxa-shared/sentry');
```
Comment (julianpoy) [inline packages/fxa-payments-server/server/lib/server.js:37]: This change seems unrelated to the PR title

Comment (sardesam) [reply packages/fxa-payments-server/server/lib/server.js:37]: I was getting a bunch of lint errors and warnings during testing that were clogging my console. This was one of them. Those items are declared but never used. I can revert in this PR. I just did it for cleanup. Please let me know!

File: libs/payments/paypal/src/lib/paypal.service.spec.ts (hunk 6 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.spec.ts:57]: Good catch! Most of it was copied over from `paypal.manager.ts`. I meant to update it like I did for `CapabilityManager` in `capability.manager.ts` but totally forgot! I'll make the change!

Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.spec.ts]: Updated. Please let me know if I need to update anything else. Ty!

File: libs/payments/currency/src/lib/currency.manager.spec.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      providers: [CurrencyManager],
+    }).compile();
+
+    mockCurrencyManager = module.get<CurrencyManager>(CurrencyManager);
+  });
+
+  it('should be defined', async () => {
+    expect(mockCurrencyManager).toBeDefined();
+    expect(mockCurrencyManager).toBeInstanceOf(CurrencyManager);
+  });
```
Comment (julianpoy) [inline libs/payments/currency/src/lib/currency.manager.spec.ts:29]: Is there a purpose for this test or is this just a placeholder?

File: libs/payments/currency/src/lib/currency.manager.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+   * Verify that provided source country and plan currency are compatible with
+   * valid currencies and countries as listed in constants
+   *
+   * @param currency Currency of customer
+   * @param country Country of customer
+   * @returns True if currency is compatible with country, else throws error
+   */
+  assertCurrencyCompatibleWithCountry(
+    currency: string | null | undefined,
+    country: string | null | undefined
```
Comment (julianpoy) [inline libs/payments/currency/src/lib/currency.manager.ts:32]: This taking both null and undefined as arguments is somewhat odd - why does this need to take both? Shouldn't currency and country both be required? The first thing you do in the method is throw an error if either aren't provided, so why not require them outright?

File: libs/payments/currency/src/lib/currency.manager.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  ): void {
+    if (!currency) throw new CurrencyCodeInvalidError(currency);
+    if (!country) throw new CountryCodeInvalidError(country);
+
+    const currencyToCountryMap: Record<string, string[]> = {
+      ...CURRENCIES_TO_COUNTRIES,
+    };
```
Comment (julianpoy) [inline libs/payments/currency/src/lib/currency.manager.ts:39]: [non-blocking] this object copy is a little deceptive. Since it's not deep copied, someone might think that this map is safe to mutate, but the array of countries is still the global array reference from the constants file - if modified, it would modify it globally.

File: libs/payments/paypal/src/lib/paypal.manager.ts [hunk 11]
```diff
@@ -35,6 +37,18 @@ export class PayPalManager {
...
   }
 
+  /**
+   * Create billing agreement using the ExpressCheckout token.
+   *
+   * If the call to PayPal fails, a PayPalClientError will be thrown.
+   *
+   */
+  async createBillingAgreement(token: string): Promise<string> {
+    const options: CreateBillingAgreementOptions = { token };
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.manager.ts:47]: You're effectively typecasting the object to this type by doing this style of definition.

It would be safer to do:

```
    const response = await this.client.createBillingAgreement({
      token,
    });
```

File: libs/payments/paypal/src/lib/paypal.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      const token = faker.string.sample();
+      const currency = faker.string.sample();
+      const options = { uid, token, currency };
+
+      mockPayPalService.createBillingAgreement = jest
+        .fn()
+        .mockResolvedValueOnce(mockNewPaypalCustomer);
+
+      const result = await mockPayPalService.createBillingAgreement(options);
+      expect(result).toEqual(mockNewPaypalCustomer);
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.spec.ts:60]: By mocking createBillingAgreement on line 55, you've made this test a non-op. Your lines 59 and 60 don't do anything, and the code under test (in the library) is never called or tested.

File: libs/payments/paypal/src/lib/paypal.service.ts [hunk 13]
```diff
@@ -4,8 +4,48 @@
...
+
+  /**
+   * Create and verify a billing agreement is funded from the appropriate
+   * country given the currency of the billing agreement.
+   */
+  async createBillingAgreement(options: {
+    uid: string;
+    token: string;
+    currency: string;
+  }) {
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.ts:27]: Since none of these are optional, these should each be independent arguments.

File: libs/payments/currency/src/lib/currency.manager.ts (hunk 9 above)
Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.ts:32]: Will update.

File: libs/payments/paypal/src/lib/paypal.service.ts [hunk 14]
```diff
@@ -4,8 +4,48 @@
...
+    const country = agreementDetails.countryCode;
+
+    this.currencyManager.assertCurrencyCompatibleWithCountry(currency, country);
+
+    const customerDetails: CreatePaypalCustomer = {
+      uid: uid,
+      billingAgreementId: billingAgreementId,
+      status: 'active',
+      endedAt: null,
+    };
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.ts:47]: You're effectively typecasting the object to this type by doing this style of definition.

It would be safer to do:

```
this.paypalCustomerManager.createPaypalCustomer({
  uid: uid,
  billingAgreementId: billingAgreementId,
  status: 'active',
  endedAt: null,
});
```

File: packages/fxa-payments-server/server/lib/server.js (hunk 7 above)
Comment (julianpoy) [reply packages/fxa-payments-server/server/lib/server.js:37]: Interesting that these aren't failing in main. Are you running your lint against just the code you changed, or against the whole project?

Comment (sardesam) [reply packages/fxa-payments-server/server/lib/server.js:37]: This was a while back so I can't quite remember. Considering it's payments-server, I'm guessing it was the whole project or just that package. 🤔

File: libs/payments/currency/src/lib/currency.manager.ts (hunk 10 above)
Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.ts:36]: I was getting type errors, hence why I opted to do it this way. But I can totally update. Let's discuss this afternoon. 👍

File: libs/payments/paypal/src/lib/paypal.manager.ts (hunk 11 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.manager.ts]: Updated!

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 14 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts:47]: Updated!

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 13 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts:27]: Updated!

File: libs/payments/currency/src/lib/currency.manager.ts (hunk 10 above)
Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.ts]: Updated as per our Slack conversation! Thank you for the suggested change!

File: libs/payments/currency/src/lib/currency.manager.spec.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+      );
+
+      expect(
+        mockCurrencyManager.assertCurrencyCompatibleWithCountry
+      ).toHaveBeenCalledTimes(1);
+
+      expect(
+        mockCurrencyManager.assertCurrencyCompatibleWithCountry
+      ).toHaveBeenCalledWith(validCurrency, validCountry);
+    });
```
[the remark this answers is not in the document]
Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.spec.ts]: Fixed as per our Slack conversation!

File: libs/payments/paypal/src/lib/factories.ts [hunk 16]
```diff
@@ -62,6 +63,14 @@ export const NVPSetExpressCheckoutResponseFactory = (
   ...override,
 });
 
+export const NVPCreateBillingAgreementResponseFactory = (
+  override?: Partial<NVPCreateBillingAgreementResponse>
+): NVPCreateBillingAgreementResponse => ({
+  ...NVPResponseFactory(),
+  BILLINGAGREEMENTID: faker.string.alphanumeric(),
```
[the remark this answers is not in the document]
Comment (sardesam) [reply libs/payments/paypal/src/lib/factories.ts:70]: I believe this is correct and follows the same as other factories in the file.

File: libs/payments/paypal/src/lib/paypal.manager.in.spec.ts [hunk 17]
```diff
@@ -299,4 +323,25 @@ describe('PaypalManager', () => {
...
   });
+
+  describe('getPayPalAmountStringFromAmountInCents', () => {
+    it('returns correctly formatted string', () => {
+      const amountInCents = faker.number.int({ max: 9999999999 });
+      const result = (amountInCents / 100).toFixed(2);
+
+      expect(
+        paypalManager.getPayPalAmountStringFromAmountInCents(amountInCents)
+      ).toEqual(result.toString());
```
[the remark this answers is not in the document]
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.manager.in.spec.ts]: Good call. Updated!

File: packages/fxa-payments-server/src/components/CouponForm/index.test.tsx [hunk 18]
```diff
@@ -15,6 +15,10 @@ import {
   CouponErrorMessageType,
 } from '../../lib/errors';
 
+// eslint-disable-next-line import/first
```
[the remark this answers is not in the document]
Comment (sardesam) [reply packages/fxa-payments-server/src/components/CouponForm/index.test.tsx]: Updated!

**Comment by sardesam** [comment]
> > There are a lot of files changed in this PR that seem unrelated to the PR title
> 
> The unrelated changes are all mentioned in the description. I can update to add them to 'Other information' if that would be more helpful. Please let me know.

Moved to a separate PR as this was affecting another PR as well. See [this comment](https://github.com/mozilla/fxa/pull/16676#issuecomment-2053834347) for more details.

File: libs/payments/currency/src/lib/currency.manager.spec.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [CurrencyManager],
+    }).compile();
+
+    mockCurrencyManager = module.get<CurrencyManager>(CurrencyManager);
+  });
+
+  describe('assertCurrencyCompatibleWithCountry', () => {
+    const validCountry = faker.helpers.arrayElement(
+      Object.values(CURRENCIES_TO_COUNTRIES.USD)
```
Comment (pdehaan) [inline libs/payments/currency/src/lib/currency.manager.spec.ts:28]: nit: I think `CURRENCIES_TO_COUNTRIES.USD` is an array, so the `Object.values()` wrapper might not be needed.

```js
export const CURRENCIES_TO_COUNTRIES = {
  USD: ['US', 'GB', 'NZ', 'MY', 'SG', 'CA', 'AS', 'GU', 'MP', 'PR', 'VI'],
  EUR: ['FR', 'DE'],
};
```

Comment (sardesam) [reply libs/payments/currency/src/lib/currency.manager.spec.ts]: Good catch! Updated!

File: libs/payments/paypal/src/lib/paypal.service.ts [hunk 20]
```diff
@@ -4,8 +4,39 @@
...
+      token
+    );
+
+    const agreementDetails = await this.paypalManager.getBillingAgreement(
+      billingAgreementId
+    );
+
+    const country = agreementDetails.countryCode;
+
+    this.currencyManager.assertCurrencyCompatibleWithCountry(currency, country);
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.ts:33]: Would be good to get @bbangert's opinion on whether this is necessary anymore with the new plan structure. I'd reached out over slack and he'd mentioned perhaps no.

File: libs/payments/paypal/src/lib/paypal.service.ts [hunk 21]
```diff
@@ -4,8 +4,39 @@
...
+    private paypalManager: PayPalManager,
+    private paypalCustomerManager: PaypalCustomerManager,
+    private currencyManager: CurrencyManager
+  ) {}
+
+  /**
+   * Create and verify a billing agreement is funded from the appropriate
+   * country given the currency of the billing agreement.
+   */
+  async createBillingAgreement(uid: string, token: string, currency: string) {
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.ts:22]: This should likely be a manager method, rather than a service-level method given the structure docs that Reino has recently posted.

File: libs/payments/paypal/src/lib/paypal.service.spec.ts [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+          casing: 'lower',
+        }),
+        billingAgreementId: faker.string.hexadecimal({
+          length: 10,
+          prefix: '',
+        }),
+        status: 'active',
+        createdAt: faker.date.recent().getTime(),
+        endedAt: null,
+      };
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.service.spec.ts:97]: These types of things should come from factories rather than in-line within tests

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 21 above)
Comment (julianpoy) [reply libs/payments/paypal/src/lib/paypal.service.ts:22]: PaypalManager would be a good location for this.

File: libs/payments/paypal/src/lib/paypal.service.spec.ts (hunk 22 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.spec.ts]: Updated!

File: libs/payments/paypal/src/lib/paypal.service.ts (hunk 21 above)
Comment (sardesam) [reply libs/payments/paypal/src/lib/paypal.service.ts]: As per our Slack conversation, overwriting method in PayPalManager and updating the test suite to include providers and factories.

File: libs/payments/paypal/src/lib/paypal.manager.in.spec.ts [hunk 23]
```diff
@@ -22,56 +24,82 @@ import {
...
+        fetchActiveCustomer: jest.fn(),
+      })
+      .overrideProvider(CurrencyManager)
+      .useValue({
+        assertCurrencyCompatibleWithCountry: jest.fn(),
+      })
+      .overrideProvider(PaypalCustomerManager)
+      .useValue({
+        createPaypalCustomer: jest.fn(),
+      })
```
Comment (bbangert) [inline libs/payments/paypal/src/lib/paypal.manager.in.spec.ts:89]: I don't believe we've used this pattern anywhere else in our code-base. The other places where we've substituted values in for a provider, we do it as its done on line 54-55. Or we create the mock separately and use it, e.g.

```typescript
const mockPaypalManager = {
  cancelBillingAgreement: jest.fn(),
  ...
};
// Then in the providers
{
  provide: PayPalManager,
  useValue: mockPaypalManager
}
```

This also avoids having to do the moduleRef.get for each one, since the mock object is already defined in a reachable location.

File: libs/payments/paypal/src/lib/paypal.manager.ts [hunk 24]
```diff
@@ -166,12 +177,31 @@ export class PayPalManager {
...
-      );
-      return await this.processNonZeroInvoice(customer, invoice);
+      await this.processZeroInvoice(invoice.id);
+      return;
+    }
+
+    const customer = await this.stripeManager.fetchActiveCustomer(
+      invoice.customer
+    );
+    await this.processNonZeroInvoice(customer, invoice);
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.manager.ts:187]: Went ahead and did this because the return values/typings were a little strange here -- not entirely related to the topic of this PR but polish.

File: libs/payments/paypal/src/lib/paypal.manager.ts [hunk 25]
```diff
@@ -24,18 +25,6 @@ export class PayPalManager {
...
-   *
-   * If the call to PayPal fails, a PayPalClientError will be thrown.
-   */
-  public async createBillingAgreement(token: string): Promise<string> {
-    const response = await this.client.createBillingAgreement({
-      token,
-    });
-    return response.BILLINGAGREEMENTID;
-  }
-
```
Comment (julianpoy) [inline libs/payments/paypal/src/lib/paypal.manager.ts:38]: [note] This stub was created in my PR previously, but can be removed now

**Review by julianpoy — APPROVED** [review]
Nice! Lets get this merged
